Docs: Revise template creation manual (#6988)
swissspidy/web-stories-wp: docs/external-template-creation.md

# Creating a new template

## Where things are stored

- The **story JSON representation** for each template is stored in [`packages/templates/src/raw/`](https://github.com/google/web-stories-wp/tree/main/packages/templates/src/raw) (in the `main` branch).

> - The **SVGs** used in each template is stored in [`assets/src/edit-story/stickers/`](https://github.com/google/web-stories-wp/tree/main/assets/src/edit-story/stickers) (in the `main` branch).

Reviewer: ```suggestion
- The **SVGs** used in each template are stored in [`assets/src/edit-story/stickers/`](https://github.com/google/web-stories-wp/tree/main/assets/src/edit-story/stickers) (in the `main` branch).
```

> - The (non-SVG) **image &amp; video files** used in each template is stored in [`public/static/main/images/templates/`](https://github.com/google/web-stories-wp/tree/static-site/public/static/main/images/templates) (in the `static-site` branch, using [Git LFS](https://git-lfs.github.com/)).

Reviewer: ```suggestion
- The (non-SVG) **image &amp; video files** used in each template are stored in [`public/static/main/images/templates/`](https://github.com/google/web-stories-wp/tree/static-site/public/static/main/images/templates) (in the `static-site` branch, using [Git LFS](https://git-lfs.github.com/)).
```

## Overview

To add a new template to the editor:

1. [Engineer] Commit all SVGs used in the template to the codebase as stickers ([details](#adding-svgs-to-the-codebase-as-stickers)).
2. [Designer] Create a new story in your local WP environment and replicate the template design.
    - SVGs should already be available in the Stickers panel from step (1) &mdash; click on a sticker to insert it into the story. If you can't find the Stickers Panel, make sure the `enableStickers` experiment is turned on.
    - Upload (drag & drop) images and videos into the story to add them.
3. [Engineer] Commit all images & videos used in the template to the codebase.
    - Make sure images are not too large &mdash; full-width images should be 1080p, large images should be 720p, and small images should be 480p. See [#6485](https://github.com/google/web-stories-wp/pull/6485) as an example.
    - Make sure videos are 720p.
4. [Engineer] Get the story JSON ([details](#get-the-story-json)) and commit it to the codebase.
    - Replace image & video URLs in the JSON with the `replaceme.com` placeholder.

> 5. [Both] Verify that new template shows up in the template library and looks as expected.

Reviewer: We should add a note about naming convention here.

In current templates we use naming convention like `travel_page9_image1.jpg`, `wellbeing_page10_image4.png`, `wellbeing_page4_bg.jpg` etc.

# Creating a new template

## Where things are stored

- The **story JSON representation** for each template is stored in [`packages/templates/src/raw/`](https://github.com/google/web-stories-wp/tree/main/packages/templates/src/raw) (in the `main` branch).

> - The **SVGs** used in each template are stored in [`assets/src/edit-story/stickers/`](https://github.com/google/web-stories-wp/tree/main/assets/src/edit-story/stickers) (in the `main` branch).

Reviewer 2: Are we ditching adding new shapes all together (like our original maskable shapes)? if not, happy to add a section on how to create shapes although it's pretty involved and requires some svg editing software and running a node script to capture the high amount of precision needed to draw complex paths in a 0 -> 1 coordinate space.

Reviewer: We aren't. Since there will be some new masks in the templates, it would be beneficial to also mention how to create the shapes. If you could add a section for this, that would be great!

Reviewer 2: Awesome. I'll create a ticket for this and add it to the sprint. it's pretty involved. I'll add a PR too with the little node script we use to normalize the svg path coords with the proper precision

Reviewer 2: https://app.zenhub.com/workspaces/web-stories-5e94d3aced449034e1e9b226/issues/google/web-stories-wp/7228

- The (non-SVG) **image &amp; video files** used in each template are stored in [`public/static/main/images/templates/`](https://github.com/google/web-stories-wp/tree/static-site/public/static/main/images/templates) (in the `static-site` branch, using [Git LFS](https://git-lfs.github.com/)).

## Overview

To add a new template to the editor:

1. [Engineer] Commit all SVGs used in the template to the codebase as stickers ([details](#adding-svgs-to-the-codebase-as-stickers)).
2. [Designer] Create a new story in your local WP environment and replicate the template design.
    - SVGs should already be available in the Stickers panel from step (1) &mdash; click on a sticker to insert it into the story. If you can't find the Stickers Panel, make sure the `enableStickers` experiment is turned on.

>     - Upload (drag & drop) images and videos into the story to add them.

Reviewer 2: Should we ddo this on staging? or some other dedicated public endpoint vs local?

Reviewer: Doesn't really matter I suppose.

On a public site you could collaborate with others during the process, so that's probably preferable.

Reviewer 2: Only reason I'd advocate for doing it on a public endpoint is that a designer could just point a dev to a story vs being responsible for capturing the story data

Reviewer: Yeah I suppose that somewhat falls under collaboration :-)

We could add a paragraph like `Note: if you are working on a template with other people, it might be a good idea to create the template on a public site where you can share access.` if we want to.

```

Once you've completed this step, your new sticker should now appear with the other stickers appended to the bottom of the shapes panel if the `enableStickers` feature flag is enabled.

### Inserting stickers (SVGs) into a story

To add stickers to your template, it's as easy as going to the experiments and enabling the `enableStickers` flag. Once turned on, all available stickers in the codebase should be available for selection under our base shapes in the shapes library panel.

### Get the story JSON

> To get the raw story json, simply open up a story in the web story editor with the devtools network tab open and press `save`. You should see the the story's json show up in the network tab. In that json, right click the `story_data` field and click `Store as global variable`. Go to the devtools console and type 

Reviewer: Alternative: just grab it from the database.

```javascript
copy(temp1)
```

When you press enter the story data should now be copied to your clipboard and ready to paste where you need.

> **Note**: `Command+Shift+Option+J / Control+Shift+Alt+J` in the story editor opens up a dialog you can copy/paste raw story json to/from. However this is missing some fields `story_data` contains which are necessary for versioning the external templates.

Reviewer: We should really fix this. Would make copying the data much easier than going to the Network tab.

### Adding story JSON to the codebase as a new template

> Once you've added your template's raw story json to `packages/templates/src/raw/<template_name>.json`, you can import it into `packages/templates/src/getTemplates` and add to the function `loadTemplates(imageBaseUrl)`. It should then be accessible in `packages/templates/src/index`.

Reviewer: There's no `loadTemplates` function.

Also, the template doesn't need to be imported individually.

```suggestion
Once you've added your template's raw story json to `packages/templates/src/raw/<template_name>.json`, you can add it to the list of templates in the `getTemplates()` function residing in `packages/templates/src/getTemplates`. It will then become accessible in the function exported by `packages/templates/src/index.js`, where you will need to add a new entry for the template (including its name, description, tags, etc.).
```

## Appendix

### Filenames for images and videos

WordPress has a naming convention where if you upload an image with the same name multiple times, it adds a `-x` suffix. 

**i.e**
if I upload `some_image.png` 3 times to the WordPress media upload in the story editor, it will store those images as:

```javascript
// some_image.png
// some_image-1.png
// some_image-2.png
```

Because of this, our conversion script will strip the `-x` suffix off of the image source so it can point to the correct image stored in our repo.

> This is an important note to make when naming images for upload into a story as **a `-x` suffix will always be stripped**. This naming convention is reserved for WordPress and will break your template creation if you use it in your image naming.

Reviewer: @littlemilkstudio Are we still doing this? I never really understood why we're doing this at runtime. We should prevent this at the ESLint stage.

Reviewer 2: It's been a while since I needed to update any media on a template, but yea I believe it's still going on.

We can add a lint step here, but honestly kinda hard to parse through the ~5000 line json files. We could try out testing this in the lint step tho and just see how it fares. Agree that it'd be nicer if it works well

Reviewer: Can you point me to the code where this is being done?

Not worried about the size of the template files for these checks. We already have Jest tests to "lint" the templates. Adding one more for file names will be trivial.

Reviewer 2: > Can you point me to the code where this is being done?

Sorry, not completely following. Where what's being done? the appending of the `-n` on image file names if uploaded multiple times?

Reviewer: "our conversion script will strip the `-x` suffix off"

Where is this happening? I remember seeing this, but can't find it right now.

Reviewer 2: 🤔 I can't seem to find it either. I'm guessing we got rid of it since the last time we were in here 🤷 . Feel like if it'd be anywhere, would be in the `loadTemplate` function

```

Once you've completed this step, your new sticker should now appear with the other stickers appended to the bottom of the shapes panel if the `enableStickers` feature flag is enabled.

### Inserting stickers (SVGs) into a story

To add stickers to your template story, go to the `Experiments` page (in the left-hand sidebar under the `Stories` section) and check the `Enable Stickers` checkbox. 

Once turned on, all available stickers in the codebase should be available for selection under our base shapes in the shapes library panel.

### Get the story JSON

To get the JSON representation of a story in the editor:

1. In the editor, open the story.
2. Open `Chrome DevTools > Network`.
3. In the editor, click the "Save draft" button. 
4. You should see a POST XHR with JSON in the request payload. In that payload JSON, find and right-click the `story_data` field and click `Copy value`. 
5. The story JSON should now be copied to your clipboard and ready to paste into a new file.

> **Note**: `Command+Shift+Option+J / Control+Shift+Alt+J` in the editor opens up a dialog you can copy/paste story JSON. However, this is missing some necessary fields for versioning.

Reviewer 2: I can't recall what fields are needed that this is missing and the importance of those fields. @choumx or @swissspidy either of you mind running through them and their importance rq for me?

Reviewer 2: is it just the version # field? does our our existing migration script update this fine if there's not one present?

Reviewer: The devtools has the version field, so that can't be it. It outputs these fields:

https://github.com/google/web-stories-wp/blob/6308e30c5766e73cdfdd3cf8aca064ee3e4ca264/assets/src/edit-story/components/devTools/devTools.js#L113-L119

That's exactly what the template contains as well:

![Screenshot 2021-04-15 at 20 34 46](https://user-images.githubusercontent.com/841956/114920735-097e0180-9e2a-11eb-8e90-4dd09077c8cc.png)

----

**Note:** `current` is the currently selected page, `selection` the currently selected element(s). Both should actually **not** be set in templates. So these templates should be fixed there. That's another thing we can catch in tests/lints). ⚠️ 

**Note 2**: `globalStoryStyles` actually refers to globally set styles in the user's database. These should **not** be set for templates, and really overridden upon loading. This is quite the bug! If you create a template on site X with 1000 global styles, and someone uses that template, they will see all those 1000 global styles instead of their own. ⚠️ 

----

What I just did:

1. Click on "use template" to load the beauty template in the editor
2. Open our devtools and copy the JSON data.
3. Replace contents of `beauty.json` with JSON data
4. Run Prettier

Besides the `globalStoryStyles` data (which is clearly a bug as per above!) and the image URLs, there were **no** differences.

So the only diffs are like this:

```diff
-            "src": "__WEB_STORIES_TEMPLATE_BASE_URL__/images/templates/beauty/beauty_page1_bg-alt.png",
+            "src": "https://wp.stories.google/static/main/images/templates/beauty/beauty_page1_bg-alt.png",
```

So I don't see why the devtools don't work for this?

Certainly way easier than trying to fiddle with Chrome DevTools!

Reviewer 2: awesome! yea the popup modal is nice. 

yea we should reset all the other fields beyond `pages` and `version` . Not sure which fields (if any) we already are doing that for, but happy to create a ticket for it.

Thanks for running me through this!

Reviewer: Created #7227 for the fieldsreset.

Reviewer: tl;dr let's advocate for using the devtools popup modal

Author: Done. Kept the devtools steps as a backup/alternative for now.

### Adding story JSON to the codebase as a new template

Once you have the story JSON, several code changes are needed to add it to the list of default templates in the editor.

> 1. In [`packages/templates/src/raw/`](https://github.com/google/web-stories-wp/tree/main/packages/templates/src/raw), commit your template's story JSON as a new file e.g. `<template_name>.json`.

Reviewer 2: I think we also need to comb through the json and update all media srcs to follow:
```
"__WEB_STORIES_TEMPLATE_BASE_URL__/images/templates/..."
```

Reviewer: Yes. There's a Jest test that complains if you forget it though.

Reviewer 2: Yea that test is great. has definitely saved me a couple times

Author: Added a substep and an example.
